### PR TITLE
Consolidate MetadataProvider and MonikerProvider

### PR DESCRIPTION
**Pull request by yufeih**
#4702, #4605 , https://ceapex.visualstudio.com/Engineering/_workitems/edit/96568/

- Cache metadata in `MetadataProvider`.
- Simplify `MonikerProvider` interface to only expose one `GetFileLevelMonikers` method, cache monikers internally.
- Fix places where `Markup` is called without passing `MonikerProvider`

### CODE REVIEW

File: src/docfx/build/xref/ExternalXRefSpec.cs [hunk 1]
```diff
@@ -15,10 +15,8 @@ internal class ExternalXrefSpec : IXrefSpec
 
         Document IXrefSpec.DeclairingFile => null;
 
-        // TODO: need a lookup table of moniker definition to get Moniker from moniker name
-        // not into output for now
         [JsonIgnore]
-        public HashSet<Moniker> Monikers { get; set; } = new HashSet<Moniker>();
```
Comment (928PJY) [inline src/docfx/build/xref/ExternalXRefSpec.cs]: conflict with this [PR](https://github.com/dotnet/docfx/pull/4621)

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 2]
```diff
@@ -2,20 +2,26 @@
...
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal class MonikerProvider
     {
-        private readonly List<(Func<string, bool> glob, (string monikerRange, IEnumerable<Moniker> monikers))> _rules = new List<(Func<string, bool>, (string, IEnumerable<Moniker>))>();
+        private readonly List<(Func<string, bool> glob, (string monikerRange, IEnumerable<string> monikers))> _rules = new List<(Func<string, bool>, (string, IEnumerable<string>))>();
```
Comment (live1206) [inline src/docfx/build/moniker/MonikerProvider.cs]: why is this changed from `Moniker` to `string` again? Needed in #4621

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 3]
```diff
@@ -98,7 +98,7 @@ private JValue TransformScalar(JsonSchema schema, Document file, Context context
                         file,
                         dependencyResolver,
                         buildChild,
-                        null,
+                        context.MonikerProvider,
```
Comment (928PJY) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: >context.MonikerProvider [](start = 24, length = 23)

we haven't design the version for sdp, keep this unchanged?

Comment (yufeih) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: Versioning should be the same for both sdp and conceptual, or the markdown pipeline should be the same for sdp and conceptual.

File: src/docfx/build/moniker/MonikerProvider.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/moniker/MonikerProvider.cs]: to move and hide moniker resolve logic inside `MonikerProvider`

File: src/docfx/build/metadata/MetadataProvider.cs [hunk 4]
```diff
@@ -98,5 +108,46 @@ private static bool IsValidMetadataType(JToken token)
...
+            {
+                return LoadSchemaDocumentMetadata(_cache.LoadYamlFile(file), file.FilePath);
+            }
+
+            if (file.FilePath.EndsWith(".json", PathUtility.PathComparison))
+            {
+                return LoadSchemaDocumentMetadata(_cache.LoadJsonFile(file), file.FilePath);
+            }
+
+            return (new List<Error>(), new JObject());
```
Comment (live1206) [inline src/docfx/build/metadata/MetadataProvider.cs]:
```suggestion
            return default;
```

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 5]
```diff
@@ -28,28 +34,47 @@ public MonikerProvider(Docset docset)
...
 
-            var (monikerError, monikers) = GetFileLevelMonikers(file, metadata.MonikerRange);
-            errors.AddIfNotNull(monikerError);
+            var zoneLevelMonikers = _rangeParser.Parse(rangeString);
+            var monikers = fileLevelMonikers.Intersect(zoneLevelMonikers, StringComparer.OrdinalIgnoreCase).ToList();
 
-            return (errors, monikers);
+            if (monikers.Count == 0)
+            {
+                var error = Errors.EmptyMonikers($"No intersection between zone and file level monikers. The result of zone level range string '{rangeString}' is '{string.Join(',', zoneLevelMonikers)}', while file level monikers is '{string.Join(',', fileLevelMonikers)}'.");
```
Comment (live1206) [inline src/docfx/build/moniker/MonikerProvider.cs]: Do we have criteria when we should pass an error message and when it should be prepared within the `Error` itself?
Because with this way, it would be harder for us to gather the error messages for PM to review

Comment (928PJY) [reply src/docfx/build/moniker/MonikerProvider.cs]: this is a special case, because we want to use the same code `empty-moniker` for the config/file and file/zone empty moniker intersection, but with different message, so we pass an error message here.

Comment (live1206) [reply src/docfx/build/moniker/MonikerProvider.cs]: The message can still be prepared

File: src/docfx/build/metadata/MetadataProvider.cs [hunk 6]
```diff
@@ -35,16 +41,23 @@ public MetadataProvider(Docset docset)
...
+        }
 
-            var result = new JObject();
-            if (yamlHeader != null)
+        private (List<Error> errors, OutputModel metadata) GetMetadataCore(Document file)
+        {
+            if (file.ContentType != ContentType.Page && file.ContentType != ContentType.TableOfContents)
             {
-                JsonUtility.SetSourceInfo(result, JsonUtility.GetSourceInfo(yamlHeader));
+                return (new List<Error>(), new OutputModel());
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/metadata/MetadataProvider.cs]: default would return a null error list, and caller would need to check null before calling `AddRange`. I also think the currently error list propagation approach is not ideal, created an [issue](https://dev.azure.com/ceapex/Engineering/_workitems/edit/98621) to track.

[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/build/metadata/MetadataProvider.cs]: right

File: src/docfx/build/moniker/MonikerProvider.cs (hunk 5 above)
Comment (yufeih) [reply src/docfx/build/moniker/MonikerProvider.cs]: @928PJY will you do this refactor later?

Comment (928PJY) [reply src/docfx/build/moniker/MonikerProvider.cs]: OK, leave it to me.
